docs/building-extensions/plugins/plugin-events/content.md

```diff
@@ -0,0 +1,214 @@
+---
+sidebar_position: 1
+title: Content Plugin Events
+---
+
+Content Plugin Events
+=====================
+
+Content events are triggered during the content creation process. These events are triggered in many different components and modules - they are not specific to the `com_content` component. 
+
+This list gives a brief description of each event, what the event arguments are, and any examples of their use which are available in the Joomla Manual.
+
+<details>
+  <summary>Accessing Event Arguments</summary>
+
+To access the event arguments of an event concrete class you can use within your plugin class:
+
+```php
+use Joomla\CMS\Event\Content\ContentPrepareEvent;
+
+class MyPlugin ...
+    public function onContentPrepare(ContentPrepareEvent $event)
+    {
+        $context = $event->getArgument('context');
+        $item = $event->getArgument('item');
+        $params = $event->getArgument('params');
+        $page = $event->getArgument('page');
+        // ...
+    }
+```
+
+However, if you want your plugin to work across Joomla 4 and 5 it may be better to use:
+
+```php
+use Joomla\Event\Event;
+
+class MyPlugin ...
+    public function onContentPrepare(Event $event)
+    {
+        [$context, $item, $params, $page] = array_values($event->getArguments());
+        // ...
+    }
+```
+
+This will work for both concrete event classes and generic event classes (as described in [Joomla 4 and 5 changes](../joomla-4-and-5-changes.md)).
+</details>
+
+<details>
+  <summary>Returning Values</summary>
+
+Where a return value is expected, how this is handled differs depending upon whether the event class is a concrete class or is a generic event class. 
+To return a value `$value` in a manner which works for both use the following:
+
+```php
+use Joomla\CMS\Event\Result\ResultAwareInterface;
+...
+    if ($event instanceof ResultAwareInterface) {
+        $event->addResult($value);
+    } else {   // use GenericEvent approach
+        $result = $event->getArgument('result') ?: [];   // get the result argument from GenericEvent
+        $result[] = $value;                              // add your return value into the array
+        $event->setArgument('result', $result);  
+    }
+```
+
+</details>
+
+## onContentPrepare
+
+### Description
+
+This is the first stage in preparing content for output and is the most common point for content-orientated plugins to do their work. 
+Since the item and related parameters are passed by reference, event handlers can modify them prior to display. 
+
+### Event Arguments

```

Comment:
Sorry for confusing. 
The old parameters cannot be directly replaced with arguments. The naming is not always matches.
But we probably still need to keep both, because the old parameter names may be useful to help transition old plugins to event classes.

Example, old listener:
```php
function($context, $item, $params, $page = 0) {}
```

And new https://github.com/joomla/joomla-cms/blob/721c32cce32f6b6f64ec8b5a4ee5817e8a5b79e5/libraries/src/Event/Content/ContentPrepareEvent.php#L18-L24:
```php
// Event
$event = new ContentPrepareEvent('onContentPrepare', [
  'context' => 'com_example.example', 
  'subject' => $contentObject, 
  'params' => $params, 
  'page' => $pageNum
]);

// Listener
function(ContentPrepareEvent $event) {
  $context = $event->getContext(); // or $event->getArgument('context');
  $item = $event->getItem(); // or $event->getArgument('subject');
  $params = $event->getParams(); // or $event->getArgument('params');
  $page = $event->getPage(); // or $event->getArgument('page');
}
```

I curently not sure how to make it right. 
Maybe keep old "Parameters" section, and later add "Arguments section". And remove "Parameters" after Joomla 6 release. Or mix it somehow. Or maybe you have some better ideas :)

Reply:
I was really following https://www.dionysopoulos.me/book/plg.html#plg-forms-j4-listener-types (see section in red below this).

I've used his suggested approach and found that it has worked ok. 

As he says in that section, even after Joomla uses a concrete event class for an event, you can still have other 3rd party components raising a generic event for that same event name, and Joomla will allow that, so your plugin has to support both. If Joomla is going to forbid that in Joomla 6 then we could move to just using concrete events at that stage. Do you know if it will?

I think that if the documentation matches the names of the arguments and the order of the arguments, then I'm hoping that this approach should work. Do you know anywhere specifically where it won't work?

Nik also mentioned some code he had submitted to ease this problem. Do you know what that is? Would including that into Joomla help things?

With regard to using Parameters, I'm reluctant to recommend a way of handling the events if we know it's going to fail in the near future (eg in Joomla 6).

Reply:
Yes, you are right, it will still work in j6, I just meant "in future" :smiley:  

`CMSPlugin::registerLegacyListener()` is not deprecated, so cannot be removed in j6, and old listeners should still work there. For now only a plan to remove in j7 https://github.com/joomla/joomla-cms/pull/43395, but still under question.

> I think that if the documentation matches the names of the arguments and the order of the arguments, then I'm hoping that this approach should work.

A few things will not match for sure. Some events have a `subject` argument, which is kind of main "subject of event". Which in old event could be as `$form`, `$data`, `$item` etc. parameter.

However, I tried make an examples, mostly every event class have an example, kind of 
- https://github.com/joomla/joomla-cms/blob/721c32cce32f6b6f64ec8b5a4ee5817e8a5b79e5/libraries/src/Event/Content/ContentPrepareEvent.php#L18-L24
- https://github.com/joomla/joomla-cms/blob/721c32cce32f6b6f64ec8b5a4ee5817e8a5b79e5/libraries/src/Event/Model/AfterSaveEvent.php#L16-L20

Where it is possible to look actual arguments.

The defination for ordering, is like https://github.com/joomla/joomla-cms/blob/721c32cce32f6b6f64ec8b5a4ee5817e8a5b79e5/libraries/src/Event/Content/ContentPrepareEvent.php#L35
(it set to remove in j6, but I think it will be extended).



> Nik also mentioned some code he had submitted to ease this problem. Do you know what that is? Would including that into Joomla help things?

I do not remember, it probably was very long time ago :wink:

Reply:
I'm coming to the conclusion that we have to explain the options to readers, and give them enough information so that they can choose what's best for them. So I've expanded the collapsible sections on how to access the parameters and return values. Would you mind checking that what I've written is correct?

I'm thinking also that we should maybe change 'item' to 'subject' in the arguments as well. What do you think? In the existing documentation it's described as 'article' which is definitely not accurate.